New and improved (and FAST) RTM package

blowfish711 opened the pull request:
The PROSPECT model and its Bayesian inversion are now written in C++ with Rcpp structures. Both of these have R wrappers for slightly more versatile calling (though the native Rcpp functions can be called directly as well).

The main new dependency is Rcpp, which I added to `install.dependencies.R`. The C++ scripts also need to be able to access the `Rmath.h` header files (for base versions of d/r/p/q statistical functions), which are present under any normal R installation and are usually referenced correctly by default building within R, but require the `-I` flag if compiled independently.

The basic inversion without random effects works very well and is VERY fast, especially for individual spectra. The speed scales linearly with size of spectra because it involves a larger matrix sum in the likelihood computation. 

The random effects inversion compiles and runs without error, but tends to get stuck after ~100 iterations. I'm currently trying to figure out what's causing this issue and reading about potential alternatives/improvements to naive Metropolis-Hastings (e.g [delayed acceptance](http://arxiv.org/abs/1406.2660), [Hamiltonian Monte Carlo](http://www.mcmchandbook.net/HandbookChapter5.pdf).

blowfish711 commented:
I should add that I migrated scripts for analyzing inversion results out of here for now, since they don't relate to the package and are currently a hot mess. Also, there is a little bit of redundancy in the data files, but they are fairly small, so I don't think it should be a major issue. As we expand this package beyond just PROSPECT4 (into 5 and 5B, as well as SAIL and other RTMs), I'll try to tidy that folder up.

mdietze commented:
Alexey, these commits are way to big. Let's sit down Thursday and go through this together since I have to many question to be able to discuss via github.

dlebauer commented on modules/rtm/DESCRIPTION:
```diff
@@ -1,37 +1,29 @@
-    Davidson, Rob Kooper, Shawn Serbin
-Maintainer: David LeBauer <dlebauer@illinois.edu>
+Author: 
+    Alexey Shiklomanov, David LeBauer, Mike Dietze,
```
Author list for a pkg should include only names of people who have contributed to writing and implementing the package

blowfish711 replied:
Ok! I pulled these from the original package and added my name, but I can reduce this list to people I know contributed.

dlebauer commented on modules/rtm/DESCRIPTION:
```diff
@@ -1,37 +1,29 @@
-    snowfall,
-    doSNOW
+    testthat
 SystemRequirements: JAGS2.2.0
```
Only specify version if necessary.

blowfish711 replied:
I'm not actually sure how to use testthat or what it does. But I'll look into it.

As for the JAGS version, again, I pulled it from the original DESCRIPTION file assuming it was that way for a reason, but I can remove the version specification.

dlebauer replied:
Please see the wiki sections on:

* testing https://github.com/PecanProject/pecan/wiki/Testing
* documentation https://github.com/PecanProject/pecan/wiki/Roxygen2

dlebauer commented on modules/rtm/R/invert_prospect.R:
```diff
@@ -0,0 +1,29 @@
+## Default inversion
+
+invert_prospect <- function(spectra,
```
Why drop the documentation?

blowfish711 replied:
The old documentation was no longer valid--I've always heard that wrong documentation is worse than none at all. When I get a chance, I'll sit town and fill in all the missing documentation.

dlebauer commented on modules/rtm/man/bayes_inversion-package.Rd:
```diff
@@ -0,0 +1,39 @@
+\name{bayes_inversion-package}
```
This file looks like a template

blowfish711 replied:
It is, it was automatically generated by the Rstudio package builder. I can delete it since it doesn't seem to affect the functioning of anything.

dlebauer replied:
yes, in general, exclude template and other unmeaningful files from commits.

dlebauer commented on modules/rtm/man/rcpp_hello_world.Rd:
```diff
@@ -0,0 +1,17 @@
+\name{rcpp_hello_world}
+\alias{rcpp_hello_world}
```
Don't need to commit this file

blowfish711 replied:
Yes, this was a mistake. I'll remove this in a future commit.

dlebauer commented on modules/rtm/src/exp_int.cpp:
```diff
@@ -0,0 +1,42 @@
+    else if (k > 0.0 and k <= 4.0){
+        xx = 0.5 * k - 1.0;
+        yy=(((((((((((((((-3.60311230482612224e-13
+        *xx+3.46348526554087424e-12)*xx-2.99627399604128973e-11)
```
Is there a rationale for using such high precision?

blowfish711 replied:
These are exactly the coefficients used in previous FORTRAN scripts (i.e. I didn't come up with them). This function is a numerical approximation of the exponential integral, which propagates through the enire calculation of the PROSPECT model. I suppose I could take the time to figure out what the minimum working precision of these values was, but that seems like a waste of time if we already have these.

dlebauer replied:
> numerical approximation of the exponential integral

I don't know enough about this particular algorithm. But this sounds like a sufficient rationale (as opposed to, say, these representing some model parameter)

dlebauer commented on modules/rtm/tests/quicktest.R:
```diff
@@ -0,0 +1,20 @@
+## Quick test
```
These look more like demonstrations than tests. More suitable as vignettes. For tests, please use the testthat testing framework.

blowfish711 replied:
OK. Like I said, I have no clue how testthat works, but I'll look into it. I wrote these scripts for testing the inversion quickly (just hitting "Source" rather than subsetting the data and organizing the results manually). I'll keep these here in their present form for now because they are very useful for debugging, but eventually, I'll restructure them as vignettes and move elsewhere.

dlebauer commented:
Did you pull upstream master before committing? Did you update the documentation using `document('rtm')` and then check?

blowfish711 commented:
Yes, one of the last few commits was a pull upstream.

But no, I did not update the documentation that way because I didn't know you could do that. Is that related to "ROxygen"? I don't have any documentation for any of the functions yet, so I figured those kinds of things wouldn't really do anything. I'll look through other PEcAn packages and try to model mine after those.

dlebauer commented:
I think this is a good example of why we need a CONTRIBUTING.md file (#298). But mostly it will point to / summarize the [Development wiki](https://github.com/PecanProject/pecan/wiki/Development), which in turn summarizes and identifies [PEcAn-specific bits from Hadley's free, online book on R programming](http://r-pkgs.had.co.nz/)

Overall, this is a very good contribution - Thanks! Sorry if that doesn't always come across in my quick notes when I am reviewing code.

blowfish711 commented:
A few updates:

In the last round of commits, I added some documentation to the R scripts (there are only 3) and removed some unnecessary files, as per the suggestions of @dlebauer . When @mdietze and I meet on Thursday, we'll discuss the rest of the recommended changes and implement accordingly.

More importantly, I made substantial changes (committed regularly in relatively small pieces; I'm trying to be better about this) that make the package much more modular (as per the recommendations of @serbinsh ). The biggest piece is that the two "statistical_inversion" scripts can now be applied to any model by simply passing the name as a string, thereby externalizing the actual model code, priors, and constraints. This was accomplished by using parameter vectors instead of scalars in RTM calls, and by using function pointers to select models from a string. The package still has only the PROSPECT4 model, but this means that adding new models and inverting them should be much easier. Also, I spoke to Keith Ma and Katia Oleinik this morning and they said that the C/Rcpp framework makes it really easy to interface with Fortran code (with a few minor caveats), so we should now be able to enhance this package with more models very quickly and efficiently.
